designs/disruption-controls-by-action.md

```diff
@@ -0,0 +1,188 @@
+# Disruption Controls By Action 
+## User Scenarios 
+1. I as a user would like to ensure upgrades only occur during business hours or stricter windows. I would also like to not restrict the cost savings of consolidation when blocking upgrade via drift.
+2. As a HFT firm, I have specific hours where I need all my compute and I can't tolerate scale down requests for consolidation, but outside of operational hours its fine to scale down. 
+## Desired Behavior + Requirements 
+1. User needs to be able to define an action and a budget/budgets for that action. 
```

Comment:
>  That way as a admin we know exactly how many nodes can be churned and by what. It makes it easier to understand that this window is specific for something.

What do you think about having rather than `All` being the total amount, `All` is the budget for all actions that are not specified in the budget list that are active at that time?

```yaml
budget: 
  action: drift 
  schedule "....." 
  nodes: 10 
budget: 
  action: consolidation 
  schedule: "....."
  nodes: 10 

budget: 
  # if action is unspecifed its the default 
  schedule: "......"
  nodes: 5 
```
 
so expiration and emptiness will allow for a total of 5 and drift and consolidation will allow for a total of 10 each.

THis allows for us to maintain the current node level controls, which are valuable themselves. Having a TotalDisruptableNodes sort of design doesn't provlde the granularity I would like. 

AKS Has [MaxSurge](https://learn.microsoft.com/en-us/azure/aks/upgrade-cluster#optimize-upgrades-to-improve-performance-and-minimize-disruptions) which is what customers will currently use to reason about upgrade. I wouldn't want to compromise on a solution where a customer cannot set the number of nodes or % of nodes that will be disrupted simulatneously during the upgrade. 

Ideally we would like to see the disruption controls to map over some of these concepts we have in AKS that allow customers a better upgrade experience.

Reply:
Updated the design to include this consideration @garvinp-stripe  please let me know what you think

Reply:
Interesting, I kind of like this. Let me give it some thought